org.eclipse.m2e.profiles.ui/src/org/eclipse/m2e/profiles/ui/internal/actions/ProfileSelectionHandler.java

```diff
@@ -75,9 +77,20 @@ public class ProfileSelectionHandler extends AbstractHandler {
 
   private static final Logger log = LoggerFactory.getLogger(ProfileSelectionHandler.class);
 
+  @Inject
+  private IProfileManager profileManager;
+
+  /**
+   * 
+   */
+  public ProfileSelectionHandler() {
+    ContextInjectionFactory.inject(this, MavenProfilesUIActivator.getServiceContext());
```

Comment:
The following variant approach would avoid changes in `MavenProfilesUIActivator`.
```suggestion
    BundleContext bundleContext = FrameworkUtil.getBundle(ProfileSelectionHandler.class).getBundleContext();
    IEclipseContext serviceContext = EclipseContextFactory.getServiceContext(bundleContext);
    ContextInjectionFactory.inject(this, serviceContext);
```

AFAICT `MavenProfilesUIActivator` until now is effectively only used to hold the plug-in ID string constant, which is only used to create I status objects. If those stati are created by the static factories the entire activator could be removed.

Or instead of going the indirection via the E4-Context respectively the `ContextInjectionFactory`, why not do the following and make `profileManager` final?
```suggestion
    BundleContext bundleContext = FrameworkUtil.getBundle(ProfileSelectionHandler.class).getBundleContext();
    profileManager = bundleContext.getService(bundleContext.getServiceReference(IProfileManager.class));
```
As you said before the service objects acquired by a bundle are unget() as soon as the bundle stops and because a handler object probably also lives until that point in time ungetting it would not free the service-object much earlier.

Reply:
> AFAICT `MavenProfilesUIActivator` until now is effectively only used to hold the plug-in ID string constant, which is only used to create I status objects. If those stati are created by the static factories the entire activator could be removed.
> 
A suggestion for that is https://github.com/eclipse-m2e/m2e-core/pull/664

Reply:
`BundleContext bundleContext = FrameworkUtil.getBundle(ProfileSelectionHandler.class).getBundleContext();` misses a lot sanity checks, I already prepared a change in E4 to make this more convenient in the next dev cycle:

- https://github.com/eclipse-platform/eclipse.platform.runtime/issues/43

Reply:
`profileManager = bundleContext.getService(bundleContext.getServiceReference(IProfileManager.class));`

This also misses some sanity checks, I think we should not use such low-level API when we have some DI in place (e.g. E4, DS, ...)

Reply:
> `BundleContext bundleContext = FrameworkUtil.getBundle(ProfileSelectionHandler.class).getBundleContext();` misses a lot sanity checks, I already prepared a change in E4 to make this more convenient in the next dev cycle:
> 
>     * [Add a method EclipseContextFactory.getServiceContext(Class<?>) method eclipse-platform/eclipse.platform.runtime#43](https://github.com/eclipse-platform/eclipse.platform.runtime/issues/43)

That change is handy and the sanity checks are indeed missing. But I think it cannot happen that this class is loaded in a non OSGi environment and we can ensure the bundle is started by ensuring `ActivationPolicy: lazy` is set in this plug-ins manifest. Therefore I doubt that they will ever fail here.

> `profileManager = bundleContext.getService(bundleContext.getServiceReference(IProfileManager.class));`
> 
> This also misses some sanity checks, I think we should not use such low-level API when we have some DI in place (e.g. E4, DS, ...)

Are you referring to the null-check on the `ServiceReference` returned by `getServiceReference()`? Or is there more I oversee?
If not, we can ensure respectively be sure that a service is provided since we do that, can't we?

Reply:
I thing you know Murphys law: the [principle](https://dictionary.cambridge.org/dictionary/english/principle) that if it is [possible](https://dictionary.cambridge.org/dictionary/english/possible) for something to go [wrong](https://dictionary.cambridge.org/dictionary/english/wrong), it will go [wrong](https://dictionary.cambridge.org/dictionary/english/wrong) ... and even in the worst possible way ;-)

Reply:
You are not wrong , but I would say if we set `ActivationPolicy: lazy` and the `BundleContext` is still null here, there is some bigger problem. :)
And mainly I would not like to keep the activator (and abandon https://github.com/eclipse-m2e/m2e-core/pull/664) just to have a saver way to get the bundle-context (if you want to be very safe, you also would have to check that the BundleContext passed to the start method is not null).